backend/manager/modules/common/src/main/java/org/ovirt/engine/core/common/businessentities/storage/ImageTransferPhase.java

```diff
@@ -25,18 +27,35 @@
     private int value;
     private String description;
     private static final Map<Integer, ImageTransferPhase> valueToPhase = new HashMap<>();
+    private static final EnumMap<ImageTransferPhase, EnumSet<ImageTransferPhase>> validTransitions = new EnumMap<>(ImageTransferPhase.class);
 
     static {
         for (ImageTransferPhase phase : values()) {
             valueToPhase.put(phase.getValue(), phase);
         }
+
+        validTransitions.put(INITIALIZING, EnumSet.of(TRANSFERRING, PAUSED_SYSTEM, CANCELLED_SYSTEM));
+        validTransitions.put(TRANSFERRING, EnumSet.of(FINALIZING_SUCCESS, PAUSED_SYSTEM, CANCELLED_SYSTEM, CANCELLED_USER, PAUSED_USER));
+        validTransitions.put(RESUMING, EnumSet.of(TRANSFERRING, PAUSED_SYSTEM, CANCELLED_SYSTEM));
+        validTransitions.put(PAUSED_SYSTEM, EnumSet.of(RESUMING, CANCELLED_USER, CANCELLED_SYSTEM));
+        validTransitions.put(PAUSED_USER, EnumSet.of(RESUMING, CANCELLED_USER, CANCELLED_SYSTEM));
+        validTransitions.put(CANCELLED_USER, EnumSet.of(FINALIZING_CLEANUP));
+        validTransitions.put(CANCELLED_SYSTEM, EnumSet.of(FINALIZING_FAILURE));
+        validTransitions.put(FINALIZING_SUCCESS, EnumSet.of(FINALIZING_FAILURE, FINISHED_SUCCESS));
+        validTransitions.put(FINALIZING_FAILURE, EnumSet.of(FINISHED_FAILURE));
+        validTransitions.put(FINALIZING_CLEANUP, EnumSet.of(FINISHED_CLEANUP));
+        validTransitions.put(FINISHED_CLEANUP, EnumSet.of(FINISHED_FAILURE));
```

Comment:
Nice!

I'm not sure the valid transitions are correct, but the validator is nice.

Reply:
Well, I guessed this based on the code, I hope eventually the spec document will provide a clear understanding of which transitions are valid. For now, I guess it would be good enough to block transitions to FINALIZING_SUCCESS when a user incorrectly finalizes a transfer

Reply:
This change is very risky - if you got the valid transitions wrong, this will break
valid code in engine trying to change the phase.

Another issue is failing user request which should succeed. For example:

More than one finalize:
1. user ask to finalize succeeds, start finalize flow
2. user ask to finalize again - fail

More than one cancel:
1. user ask to cancel succeeds, start finalize flow
2. user ask to cancel again - fail

The system will be much easier to use if calling finalize() or cacnel() more than
once is allowed and does not change anything.

So this cannot be done using this validation. I think adding flags like:

- finalized
- canceled 

When we handle finalize/cancel request we can consider the flags before we 
modify the phase.

Reply:
Sure, this is a draft to showcase a possible direction for validating, and I am in no rush to merge this. I still need to add unit tests and test this extensively, but even before that, we need to complete the spec, at least transitions part.

Multiple user requests will not fail (same goes for internal phase updates), but won't change the phase. But I am not sure if that is correct in all the time, maybe in some cases the entire entity update should be discarded...

Reply:
The problematic flow I see is this:

1. User ask to finalize
2. System change phase to finalizing success
3. System change state to the next phase
4. User ask to finalize
5. System changes state to finalizing success
6. System remain stuck in finalizing success

And same for cancel.

I could reproduce this issue with virt-v2v:
1. virt-v2v ask to finalize without closing imageio socket
2. engine waits for imageio to remove the ticket
3. imageio times out, because of the open socket
4. virt-v2v request times out before engine replies
5. engine wrongly advance to the next phase after imageio timeout
6. virt-v2v cancel the transfer because of the finalize timeout
7. engine get stuck in finalizing failure

This is a combinations of bugs in virt-v2v, imageio and engine
that cannot happen now because virt-v2v does close the socket now
before calling finalize, and imageio does not time out when there
are not active requests, and will remove the ticket immediately, but
this can happen with another clients for other reasons.

To fix these issues, we must ensure that requests to finalize
or cancel are handled exactly once.